### PR TITLE
feat: add initial version

### PR DESCRIPTION
**Pull request by chrisgacsal**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | 
| License         | Apache 2.0

### What's in this PR?

Add initial version of the project.

### Checklist

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### CODE REVIEW

File: .golangci.yml
Comment (adamantal) [inline .golangci.yml:8]: What is the rationale behind only enabling certain linters? I mean if this repo would want to be compliant with the Go linter rules, this may be the time to fix them now.

File: README.md
Comment (adamantal) [inline README.md:6]: Though CC is quite fast at creating new version (so maintainers can update the latest CC version every day), but I think it would be beneficial to include the latest tested or verified CC versions in the readme.

**Review by adamantal — COMMENTED** [review]
Just finished with the first few files. 
Taking into account the size of the PR, I'll keep pushing my comments as they go so you'll have time to fix them.

File: README.md
Comment (chrisgacsal) [reply README.md:6]: Agreed. I planned to do that, but forgot about it. Will do.

File: .golangci.yml
Comment (chrisgacsal) [reply .golangci.yml:8]: By default only a subset of linters are enabled. The rest need to be explicitly be enabled. See: https://golangci-lint.run/usage/quick-start/

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (chrisgacsal) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: Agreed! Will add!

File: .golangci.yml
Comment (chrisgacsal) [reply .golangci.yml:8]: Also this way you can make sure that the expected linters are enabled even if the default set changed.

File: README.md
Comment (baluchicken) [inline README.md:82]: Should we replace this link to `http://community-banzaicloud.slack.com` ?

Comment (chrisgacsal) [reply README.md:82]: Yep. Agreed.

File: deploy/docker-compose.yml
Comment (baluchicken) [inline deploy/docker-compose.yml:42]: Should we bump this to 2.5.80?

Comment (chrisgacsal) [reply deploy/docker-compose.yml:42]: We could, although there is not AP change between the two versions. I'll bump the version but in a separate PR.

**Review by baluchicken — APPROVED** [review]
Just nits:)

File: pkg/api/rebalance.go
Comment (adamantal) [inline pkg/api/rebalance.go:116]: I haven't thought this thoroughly, but I might have a suggestion regarding this function and how to eliminate the code duplication.

Instead of this function:
```golang
func (r *RebalanceResponse) UnmarshalResponse(resp *http.Response) error {
```
You can write something like:
```golang
func UnmarshalResponse(resp *http.Response, into *TargetResponse) error {
```
where
```
type TargetResponse interface {
  SetResult(types.OptimizationResult)
  SetProgress(types.ProgressResult)
  SetError(types.APIError)
}
```
can be implemented by the `*Response` types all around the `api` package. You still have to implement these (or move these implementations to `types.GenericsResponse` maybe?), but I feel like it's less code than the current `UnmarshalResponse`. 

I am not sure this works, just wanted to come up with a solution about the multiple `UnmarshalResponse` functions.  WDYT?

File: pkg/api/train.go
Comment (adamantal) [inline pkg/api/train.go:48]: I think this is wrong. According to the description, `0` is a valid value (the current time will be used instead of zero).
Though `omitempty` is there, the Go struct int value will be initialized to zero, therefore the validation fails. 
Or am I missing something here?

**Review by adamantal — COMMENTED** [review]
Got through api package, only commenting on the non-nits

File: pkg/api/train.go
Comment (chrisgacsal) [reply pkg/api/train.go:48]: Nope. You are right. Will fix this by adding the right defaults and using constraints `Start < -1` and `End < 0` for validation.

File: pkg/api/rebalance.go
Comment (chrisgacsal) [reply pkg/api/rebalance.go:116]: Yeah. I was thinking about this as well. The type of the `.Result` attribute vary depending on the API Response. E.g. the `Rebalance` API returns a `types.OptimizationResult` while other APIs like `KafkaParitionLoad` API returns `types.PartitionLoadState`. So using a single interface would not help a lot with removing all the code duplication. I agree that there are ways to get a more optimal implementation here (groups APIs based on responses, etc), I also wanted to keep it clean by not adding lot off "magic" which could degrade the readability of the code.

I consider this as an enhancement and I think it should be a separate effort.

File: deploy/wait-for-it
Comment (chrisgacsal) [reply deploy/wait-for-it:57]: Valid. The script is copied as is. However will fix this too.

Comment (chrisgacsal) [reply deploy/wait-for-it]: Fixed.

File: pkg/api/train.go
Comment (chrisgacsal) [reply pkg/api/train.go]: Fixed.

File: pkg/types/params.go
Comment (adamantal) [inline pkg/types/params.go:21]: It has been talked before to move the `properties` package from koperator to a separate one. Seems like some of the functionalities has been reimplemented here. Wouldn't it make sense to use that package here?

File: README.md
Comment (chrisgacsal) [reply README.md]: Fixed.

File: pkg/types/topic_partition.go
Comment (adamantal) [inline pkg/types/topic_partition.go:22]:
```suggestion
```

File: deploy/docker-compose.yml
Comment (chrisgacsal) [reply deploy/docker-compose.yml:42]: Fixed.

**Review by adamantal — COMMENTED** [review]
Got through the whole `pkg` dir, and it looks good to me.
Lots of manually typed enums though, it must have been ages to do this. 🥇 Good job!

I haven't had time to get through the integration test part, but if you fix these items, I don't want to hold this PR back any longer.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (chrisgacsal) [reply .github/PULL_REQUEST_TEMPLATE.md:18]: Fixed.

File: README.md
Comment (chrisgacsal) [reply README.md:82]: Fixed.

File: pkg/client/config.go
Comment (chrisgacsal) [reply pkg/client/config.go:23]: The reason why I decided no to do that to keep the dependencies for this lib as minimal as possible. Since the configuration is not complex (so far), I do not think we need external lib for managing configuration.

File: pkg/client/endpoints.go
Comment (chrisgacsal) [reply pkg/client/endpoints.go:34]: I will remove this link. It was just a reminder for me.

File: pkg/client/endpoints.go
Comment (chrisgacsal) [reply pkg/client/endpoints.go:140]: I'll do that in a separate PR.

File: pkg/types/params.go
Comment (chrisgacsal) [reply pkg/types/params.go:21]: The `properties` pkg is not relevant here, therefore I prefer not to pull it in. Only the struct tag handling logic is duplicated between the `properties` pkg and this one.

File: pkg/types/topic_partition.go
Comment (chrisgacsal) [reply pkg/types/topic_partition.go:22]: Will fix this.

**Review by baluchicken — APPROVED** [review]

File: pkg/client/endpoints.go
Comment (chrisgacsal) [reply pkg/client/endpoints.go]: Fixed.

File: pkg/types/topic_partition.go
Comment (chrisgacsal) [reply pkg/types/topic_partition.go]: Fixed.

File: README.md
Comment (chrisgacsal) [reply README.md:8]: Fixed.

**Review by baluchicken — APPROVED** [review]

File: pkg/api/rebalance.go
Comment (adamantal) [reply pkg/api/rebalance.go:116]: Understood

File: pkg/client/config.go
Comment (adamantal) [reply pkg/client/config.go:23]: Sounds reasonable. If there were that many config options, we would start to use one anyways.

File: pkg/types/params.go
Comment (adamantal) [reply pkg/types/params.go:21]: Discussed offline. Good to go

**Review by adamantal — APPROVED** [review]

**Pull request merged**
